### PR TITLE
FIPS: digest checking for FIPS provider

### PR DESCRIPTION
**Pull request by paulidale**
Add a configuration backed indicator for the signature algorithm digest checks.

Bar XOF digests for algorithms where they are not supported (both for FIPS and default providers).

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: providers/implementations/signature/rsa_sig.c
Comment (slontis) [inline providers/implementations/signature/rsa_sig.c:331]: Unless we support this it will probably just fall over later??

File: util/mk-fipsmodule-cnf.pl
Comment (slontis) [inline util/mk-fipsmodule-cnf.pl:18]: It probably doesnt matter, but I would have left them as 2 seperate items..

File: test/recipes/03-test_fipsinstall.t
Comment (slontis) [inline test/recipes/03-test_fipsinstall.t:30]: Why is this needed here?

File: providers/implementations/signature/rsa_sig.c
Comment (paulidale) [reply providers/implementations/signature/rsa_sig.c:331]: It will, but it's on NIST's permitted list.  Thus we should make a token effort to support it.

File: util/mk-fipsmodule-cnf.pl
Comment (paulidale) [reply util/mk-fipsmodule-cnf.pl:18]: Meh.  We want them all enabled by default.

File: test/recipes/03-test_fipsinstall.t
Comment (paulidale) [reply test/recipes/03-test_fipsinstall.t:30]: It's an extra options when installing the FIPS provider.
This makes an effort to check that the option is accepted.

Comment (slontis) [reply test/recipes/03-test_fipsinstall.t:30]: Are all options supposed to be here in general (Just a general question, not related to this PR)

File: providers/implementations/signature/rsa_sig.c
Comment (slontis) [reply providers/implementations/signature/rsa_sig.c:331]: I think we should support only what we allow? Not all algorithms need to be supported.

File: test/recipes/03-test_fipsinstall.t
Comment (paulidale) [reply test/recipes/03-test_fipsinstall.t:30]: Dunno.  It seemed like a reasonable idea to check that the option is accepted.

File: providers/implementations/signature/rsa_sig.c
Comment (paulidale) [reply providers/implementations/signature/rsa_sig.c]: Updated.

File: providers/implementations/signature/dsa_sig.c
Comment (slontis) [inline providers/implementations/signature/dsa_sig.c:172]: blankline?

**Review by slontis — APPROVED** [review]

**Comment by paulidale** [comment]
Ping @openssl/committers for second review.

**Comment by paulidale** [comment]
Rebased to avoid trivial conflict.  Re-ping for second review @openssl/committers

File: apps/fipsinstall.c
Comment (tmshort) [inline apps/fipsinstall.c:332]: Was this simply incorrect originally?

File: providers/fips/fipsprov.c
Comment (tmshort) [inline providers/fips/fipsprov.c:203]: Why `[28]`? Only one additional parameter was added. Shouldn't this be `[27]`?

File: providers/implementations/signature/dsa_sig.c
Comment (tmshort) [inline providers/implementations/signature/dsa_sig.c:172]: Is the `&` operator necessary here, and elsewhere? `FIPS_fips_signature_digest_check` is already a function, so should already be a pointer?

Comment (slontis) [reply providers/implementations/signature/dsa_sig.c:172]: There is going to be a pass over these after all the changes are in, to change the names to something more meaningful, this should be resolved at that time.

File: apps/fipsinstall.c
Comment (paulidale) [reply apps/fipsinstall.c:332]: Yes, although both just happen to be defined to the same string.

File: providers/implementations/signature/dsa_sig.c
Comment (paulidale) [reply providers/implementations/signature/dsa_sig.c:172]: The `&` is just a habit of mine.  I find it more readable and it highlights that a function pointer is being passed.
The coding style is mum on including it or not.

File: providers/fips/fipsprov.c
Comment (paulidale) [reply providers/fips/fipsprov.c:203]: I was attempting to avoid a conflict with one of my other in flight PRs.

Yes, it should be 27 but since we get a conflict ever merge, it will end up correct.
I've been tempted to submit a PR converting this to a stack of params.  More than once.

File: providers/implementations/signature/dsa_sig.c
Comment (paulidale) [reply providers/implementations/signature/dsa_sig.c:172]: I'm not sure we need to rename these -- they are inside the FIPS provider which can only be built as a shared object and the namespace is kept completely internal.  I.e. they cannot pollute the calling application's namespace.

File: providers/fips/fipsprov.c
Comment (paulidale) [reply providers/fips/fipsprov.c:203]: After merging #24120, it now needs to be 28 :grin:

File: providers/implementations/signature/dsa_sig.c
Comment (slontis) [reply providers/implementations/signature/dsa_sig.c:172]: It was more about putting 'config' in the name

**Comment by paulidale** [comment]
Re-ping @openssl/committers for second review please.

File: providers/implementations/signature/sm2_sig.c
Comment (beldmit) [inline providers/implementations/signature/sm2_sig.c:105]: Does this check make sense for SM2 at all?

**Review by beldmit — APPROVED** [review]
LGTM with or without SM2 changes

File: providers/implementations/signature/sm2_sig.c
Comment (paulidale) [reply providers/implementations/signature/sm2_sig.c:105]: I'm not aware that SM2 supports XOF digests.  I could easily be mistaken, the standards are all Chinese to me :grin:

I'm pretty sure the code doesn't support them currently.

**Comment by t8m** [comment]
Needs another rebase

**Comment by paulidale** [comment]
Test failures are relevant, I'll investigate tomorrow.

File: providers/implementations/signature/dsa_sig.c
Comment (tmshort) [reply providers/implementations/signature/dsa_sig.c:172]: Perhaps the coding standard should be updated to resolve this ambiguity.

**Comment by paulidale** [comment]
Test failures are for the FIPS provider in this PR against older branches.  I think it's going to require another PR against the old branches to address.  Sigh.

The minimum would be updating the test cases, but we might need code changes in evp_test (I'm not sure if FIPSversion and AvailableIn stack or not).

**Comment by paulidale** [comment]
#25133 fixes the cross version tests that fail with this PR.  They should be merged together.

**Comment by paulidale** [comment]
@slontis @beldmit re-review please?

It's just been rebased a couple of times, the test changes that were introduced in [274b6ae](https://github.com/openssl/openssl/pull/25020/commits/274b6ae1058942f678f47ab5ea6203004fe9bc79) have been reverted.

The real test fixes are in #25133 and it is expected that evp_test will fail here for the cross compatibility checks.

**Comment by paulidale** [comment]
@beldmit re-approval please?  It's back to what you approved previously with rebasing to maintain a clean merge to master.

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by beldmit** [comment]
@paulidale sorry, lost it in inbox yesterday :(

**Comment by beldmit** [comment]
What's going on with provider-compat tests?

**Comment by paulidale** [comment]
Okay, those test here **must** fail until #25133 is merged.  It is essentially a change of behaviour between what was before and the current FIPS provider.  There is literally nothing that can be done to rectify it beyond updating the older branches.

> Testing 3.0 against the 3.4 FIPS provider will result in failures due to NIST moving the goal posts.  If we update the 3.0 tests to account for this, they will pass.

I intend to merge #25133 before this which should resolve things.

**Comment by beldmit** [comment]
Let me rerun the tests after #25133 is merged and I will approve after that

**Comment by paulidale** [comment]
Tests rerunning...

**Comment by paulidale** [comment]
Cross compatibility CI error is between this PR and master -- i.e. expected.

**Review by beldmit — APPROVED** [review]
LGTM, the test failure should disappear on merging

**Comment by paulidale** [comment]
Merged, thanks for the feedback and your patience.

**Pull request closed without merging**
